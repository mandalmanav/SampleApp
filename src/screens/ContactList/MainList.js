import React, { useEffect } from 'react';
import { View, Text, SectionList, StyleSheet, FlatList, Pressable } from 'react-native'
import { useSelector, useDispatch } from 'react-redux'
import { getAllContactsAction } from '../../store/actions/contacts';
import HighlightText from '@sanar/react-native-highlight-text';
import { SET_SELECTED } from '../../store/constants';
const Item = ({ title, isLast, searchTerm }) => (
    <View style={styles.item}>

        <Text testID="contact" style={isLast ? { ...styles.title, ...styles.border0 } : { ...styles.title, ...styles.border1 }}>
            <HighlightText
                highlightStyle={{ fontWeight:'bold',backgroundColor:'yellow' }}
                searchWords={[searchTerm]}
                textToHighlight={title}
            />
        </Text>
    </View>
);
export default ({  navigate }) => {
    const dispatch = useDispatch();
    let { contacts ,searchTerm} = useSelector(state => state)
    useEffect(() => {
        getAllContactsAction(dispatch)
    }, [])
    let groupList = [
        {
            title: 'A',
            data: []
        }, {
            title: 'B',
            data: []
        }, {
            title: 'C',
            data: []
        }, {
            title: 'D',
            data: []
        }, {
            title: 'E',
            data: []
        }, {
            title: 'F',
            data: []
        }, {
            title: 'G',
            data: []
        }, {
            title: 'H',
            data: []
        }, {
            title: 'I',
            data: []
        }, {
            title: 'J',
            data: []
        }, {
            title: 'K',
            data: []
        }, {
            title: 'L',
            data: []
        }, {
            title: 'M',
            data: []
        }, {
            title: 'N',
            data: []
        }, {
            title: 'O',
            data: []
        }, {
            title: 'P',
            data: []
        }, {
            title: 'Q',
            data: []
        }, {
            title: 'R',
            data: []
        }, {
            title: 'S',
            data: []
        }, {
            title: 'T',
            data: []
        }, {
            title: 'U',
            data: []
        }, {
            title: 'V',
            data: []
        }, {
            title: 'W',
            data: []
        }, {
            title: 'X',
            data: []
        }, {
            title: 'Y',
            data: []
        }, {
            title: 'Z',
            data: []
        }
    ]
    let filteredList = []
    if (searchTerm.trim() !== "") {
        filteredList = contacts.filter(item =>( item.firstName&&item.firstName.toLowerCase().indexOf(searchTerm.toLowerCase()) != -1)|| (item.lastName&&item.lastName.toLowerCase().indexOf(searchTerm.toLowerCase()) != -1))
    }
    else {
        filteredList = contacts;
    }
    let ungroupContacts = {
        title: '#',
        data: []
    }
    let arrayOfInitials = groupList.map(group => group.title.toLowerCase())

    let groupedContacts = groupList.map(item => {
        filteredList.map(contact => {

            if (contact.firstName&&contact.firstName.toLowerCase().startsWith(item.title.toLowerCase())) {
                item.data.push(contact)

            }


        })


        return item
    })
    filteredList.map(contact => {

        if (arrayOfInitials.indexOf(contact.firstName&&contact.firstName.toLowerCase().charAt(0)) == -1) {
            ungroupContacts.data.push(contact)
        }

    })
    groupedContacts.push(ungroupContacts)
    groupedContacts = groupedContacts.filter(group => group.data.length)
    const renderItem = ({ item, index, section, ...rest }) => (
        
        <Pressable onPress={() => {
            dispatch({
                type:SET_SELECTED,
                payload:item
            })
            navigate('ContactDetails', { contact: item })
        }}>
            {
                section ? <Item title={`${item.firstName} ${item.lastName}`} isLast={index === section.data.length - 1} /> : <Item searchTerm={searchTerm} title={`${item.firstName} ${item.lastName}`} isLast={index === filteredList.length - 1} />
            }
        </Pressable>
    );
    return <View style={styles.content}>
        {
            searchTerm.trim() == "" ? <SectionList
                sections={groupedContacts}
                keyExtractor={(item, index) => item + index}
                renderItem={
                    renderItem
                }
                renderSectionHeader={({ section: { title } }) => (
                    <Text style={styles.header}>{title}</Text>
                )}
            /> : <FlatList
                data={filteredList}
                renderItem={renderItem}
                keyExtractor={item => String(item.id)
                }
            />

        }

    </View>
}

const styles = StyleSheet.create({
    header: {
        backgroundColor: '#f0f0f0',
        height: 24,
        textAlignVertical: 'center',
        paddingLeft: 4,
        fontWeight: "bold"
    },
    item: {
        padding: 4,
        paddingTop: 8,
    },
    title: {
        paddingBottom: 8
    },
    border0: {
        borderBottomWidth: 0
    },
    border1: {
        borderBottomWidth: .3,
        borderBottomColor: '#e7e7e7'
    },
    content: {
        paddingLeft: 8,
        paddingRight: 8,
        flex: 1,
        marginTop: 12
    }

})