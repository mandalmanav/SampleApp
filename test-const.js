export const CONTACTS=[{
    firstName:'Manav',
    lastName:'Mandal',
    image:'',
    mobile:'9999999999',
    id:'q',
    countryCode:'+91'
},{
    firstName:'Aman',
    lastName:'Vatsa',
    image:'',
    mobile:'9999999999',
    id:'r',
    countryCode:'+91'
}]

export const SEARCH_TERM='am'