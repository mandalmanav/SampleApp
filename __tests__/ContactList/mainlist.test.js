
import { render, fireEvent } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import MainList from '../../src/screens/ContactList/MainList'
import { CONTACTS, SEARCH_TERM } from '../../test-const';


describe('Test Contact List', () => {
    
    let initialState = { contacts: CONTACTS, searchTerm: '' }
    const mockStore = configureStore()
    let store, wrapper
    store = mockStore(initialState)
    it('Showing full list',()=>{
        const {getAllByTestId} = render(<Provider store={store}><MainList /></Provider>)
        expect(getAllByTestId('contact').length).toBe(CONTACTS.length)
    })
    it('List is filtered',()=>{
        initialState = { contacts: CONTACTS, searchTerm: CONTACTS[0].firstName }
        store = mockStore(initialState)
        const {getAllByTestId} = render(<Provider store={store}><MainList /></Provider>)
        expect(getAllByTestId('contact').length).toBe(1)
    })
    it('Match Snapshot', () => {
        const component = render(<Provider store={store}><MainList /></Provider>)
        expect(component).toMatchSnapshot()
    })
})