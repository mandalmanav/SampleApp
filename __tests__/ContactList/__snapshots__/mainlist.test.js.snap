// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Contact List Match Snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 12,
      "paddingLeft": 8,
      "paddingRight": 8,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "countryCode": "+91",
          "firstName": "Manav",
          "id": "q",
          "image": "",
          "lastName": "Mandal",
          "mobile": "9999999999",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              Object {
                "padding": 4,
                "paddingTop": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "borderBottomWidth": 0,
                  "paddingBottom": 8,
                }
              }
              testID="contact"
            >
              <Text>
                <Text
                  style={
                    Object {
                      "backgroundColor": "yellow",
                      "fontWeight": "bold",
                    }
                  }
                >
                  Manav
                </Text>
                 Mandal
              </Text>
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
